### DISCUSSION TITLE
Cross-compilation to macos aarch64-apple-darwin target using Maturin Docker

### DISCUSSION THREAD

**Discussion by mikwieczorek** [PyO3/maturin#1768] [Q&A] [answered]
I am attempting to cross-compile Python Rust bindings for macOS aarch64-apple-darwin using maturin. I've tried using the maturin Docker image, but it fails with an error related to the Python interpreter. I've also looked into using osxcross for cross-compilation, but I'm not sure how to integrate it with maturin.

Based on the GitHub issue [#375](https://github.com/PyO3/maturin/issues/375) and the comments from @messense mentioning the `--zig` option and [python-build-standalone](https://github.com/indygreg/python-build-standalone) (not sure how to use it), I'm seeking a more streamlined solution.

I followed the tips from the aforementioned Issue and docs from PyO3 [here](https://pyo3.rs/v0.19.2/building_and_distribution.html#advanced-config-files). I tried using latest maturin docker image, but I couldn't make it work and compile my project to `aarch64-apple-darwin`.

When I run:
```
  export PYO3_PYTHON=/python/bin/python3.8
  export PYO3_CROSS=1
  export PYO3_CROSS_LIB_DIR=/python/lib
  export PYO3_CROSS_PYTHON_VERSION=3.8
  export PYO3_CROSS_PYTHON_IMPLEMENTATION=CPython
  maturin build --interpreter $(which python3.8)--release --target aarch64-apple-darwin;
```
( I tired with and without additional flags in Cargo.toml:  `rustflags = [ "-C", "link-arg=-undefined",  "-C", "link-arg=dynamic_lookup",]`

I got 
```
  error: linking with `cc` failed: exit status: 1 cc: error: arm64: No such file or directory
  cc: error: dynamic_lookup: No such file or directory
  cc: error: unrecognized command line option '-arch'; did you mean '-march='?
```

When I added `--zig` flag (and installed ziglang) then the error from build changed to: `note: error: unsupported linker arg: -exported_symbols_list`

I also tried add maturin on top of osxcross and playing with `CC` and `CXX` variables but to no effect. 

Finally I tried using `abi3-py38` feature to avoid providing path to macOS complied python lib, but this also didn't work.

Questions

1. Is there a straightforward way to cross-compile to macOS aarch64-apple-darwin using the maturin Docker image?
2. Do I need to use osxcross with maturin for cross-compilation?
3. Could you provide a minimal working example for either of the above methods?

I would greatly appreciate any guidance or examples to help resolve this issue. Thank you!

**Comment by messense** [comment, marked as answer]
> 1. Is there a straightforward way to cross-compile to macOS aarch64-apple-darwin using the maturin Docker image?

No, the docker image is for native compilation only.

> 2\. Do I need to use osxcross with maturin for cross-compilation?

No, but you can use it, there is nothing in maturin that prevents you from using a custom toolchain. You can use zig, but you should probably use zig 0.10.1 instead of 0.11.x because zig 0.11.x errors on unsuppported linker flags (instead of warning like in 0.10.1).

> 3\. Could you provide a minimal working example for either of the above methods?

Sorry I don't have the free time to do that.

It's hard to guess what's the real problem without a repro, so please provide a minimal repro. If you can't provide a repro in public for reasons, I'm up for paid private support if you want.

**Comment by mikwieczorek** [comment]
Thank you @messense you tip on using zig 0.10.1.post1 solved the problem. 
Additionally I had to prepare some info following format in `lib/_sysconfigdata*.py` file to allow proper extensions etc. for built wheels. Setting `PYO3_CROSS_LIB_DIR` to this prepared '_sysconfigdata*.py` resulted in successful build.
